tests/test_TFUtil.py

```diff
@@ -469,6 +469,30 @@ def test_Data_copy_compatible_to_batch_feature_is_dynamic():
   assert_equal(t.get_size_dim_tag(0), dec.get_time_dim_tag())
 
 
+def test_Data_get_common_data_modified_time_dim():
+  d1 = Data(name='t', shape=(None, 32, 128), dtype='float32', auto_create_placeholders=True)
+  d2 = Data(name='r', shape=(None, 32, 128), dtype='float32', auto_create_placeholders=True)
+  # simulating the changes in the time dim
+  d1.size_placeholder[0] -= 1
+  d2.size_placeholder[0] -= 1
+
+  from TFUtil import DimensionTag
+  # doing tagging a la SliceLayer for d1.size_placeholder
+  if not DimensionTag.get_tag_from_size_tensor(d1.size_placeholder[0]):
+    tag = DimensionTag(
+      description="slice1",
+      kind=DimensionTag.Types.Spatial)
+    tag.set_tag_on_size_tensor(d1.size_placeholder[0])
+
+  # WindowLayer does no tagging on size_placeholder
+  # => d2.size_placeholder untagged
```

Comment:
Remove this comment. This is just about `TFUtil`.
Also, the comment is wrong. `d2.size_placeholder` is tagged (when you remove the `d2.size_placeholder[0] -= 1`).
What case are you actually trying to produce here? That `d1` and `d2` have different time dim tags? You already have that from the very beginning.
Or that `d1.size_placeholder` is tagged but `d2.size_placeholder` is not tagged?

Btw, if you are suggesting that `WindowLayer` does not correctly set the time dim tag, then you should add a test case which shows and demonstrates and checks for exactly that: I.e. just using and testing `WindowLayer` (nothing else; keep focused in a test case) (this would go to `test_TFNetworkLayer.py` again).

Reply:
> Or that `d1.size_placeholder` is tagged but `d2.size_placeholder` is not tagged

You asked to reproduce the error within TFUtil.py. 

> Can you try to reproduce this problem just by using Data (or other things from TFUtil), and create a test case in test_TFUtil.py?

Here's the case.

Reply:
> Btw, if you are suggesting that `WindowLayer` does not correctly set the time dim tag, then you should add a test case which shows and demonstrates and checks for exactly that: I.e. just using and testing `WindowLayer` (nothing else; keep focused in a test case) (this would go to `test_TFNetworkLayer.py` again).

From the very beginning of the discussion I suggested, that tagging doesn't work properly for WindowLayer and SliceLayer.

Reply:
I'm not sure what you are saying. Please answer my comment / question.

Reply:
> Btw, if you are suggesting that `WindowLayer` does not correctly set the time dim tag, then you should add a test case which shows and demonstrates and checks for exactly that: I.e. just using and testing `WindowLayer` (nothing else; keep focused in a test case) (this would go to `test_TFNetworkLayer.py` again).

From the very beginning I suggested that the tagging system doesn't work for some layer. In particular, SliceLayer assigns one tag, Window layer assigns no tag. And as a result the outputs of these two layers can't be processed in the CombineLayer. So, I do the test on this particular situation.

Reply:
> From the very beginning of the discussion I suggested, that tagging doesn't work properly for WindowLayer and SliceLayer.

Where? Make a very specific test case for exactly / only that. One for `WindowLayer` only. Another one for `SliceLayer` only. Just show directly where sth is wrong. Keep focused in a test case. Keep it simple.

Also, you did not answer/addressed my comments here:

Remove this comment. This is just about TFUtil.
Also, the comment is wrong. d2.size_placeholder is tagged (when you remove the d2.size_placeholder[0] -= 1).

What case are you actually trying to produce here?
That d1 and d2 have different time dim tags? You already have that from the very beginning.
Or that d1.size_placeholder is tagged but d2.size_placeholder is not tagged?
What case exactly? (Don't just answer "the error I have". Say *what error*, *what situation exactly*.)

Reply:
> > From the very beginning of the discussion I suggested, that tagging doesn't work properly for WindowLayer and SliceLayer.
> 
> Where? Make a very specific test case for exactly / only that. One for `WindowLayer` only. Another one for `SliceLayer` only. Just show directly where sth is wrong. Keep focused in a test case. Keep it simple.

`WindowLayer` and `SliceLayer` work ok when separate. The main issue is that the tags on the placeholders are different, when processed by different layers. And they can't be processed by the `CombineLayer`. I think, one of the solution, is to make `CombineLayer` less discriminative. Maybe by modifying `Data.get_common_data()`, s.t. `common_data` has a single time_dim even if the inputs have different tags on it.

> Remove this comment. This is just about TFUtil.
> Also, the comment is wrong. d2.size_placeholder is tagged (when you remove the d2.size_placeholder[0] -= 1).
I removed `d2.size_placeholder[0] -= 1`


> What case are you actually trying to produce here?
> That d1 and d2 have different time dim tags? You already have that from the very beginning.
> Or that d1.size_placeholder is tagged but d2.size_placeholder is not tagged?
> What case exactly? (Don't just answer "the error I have". Say _what error_, _what situation exactly_.)

I reproduce the test s.t. two Data instances have different tags for the time dim. Then, I create a `common_data`. Then make a copy for d1 and d2. Then compare the shapes. The same procedure as in the `CombineLayer`. Because you asked to reproduce the problem by using the `Data`.

Reply:
> `WindowLayer` and `SliceLayer` work ok when separate. The main issue is that the tags on the placeholders are different, when processed by different layers.

This is a contradiction. I'm not sure what you are saying. The tags are the same before these layers, and then become different? (If you mean that, why are you not directly saying and showing that?) And this is wrong, or not? If this is wrong, you can clearly just make an individual test for that (checking that the time dim is the same as it was before the individual layer). If it is correct, then I don't see where the problem is, because CombineLayer is also correct. I.e. this is a contradiction.

> And they can't be processed by the `CombineLayer`. I think, one of the solution, is to make `CombineLayer` less discriminative.

No, I explained already, many times now, the behavior of `CombineLayer` (or more specifically, `Data.get_common_data`) is perfectly correct (for all the cases which you have shown so far).

> > What case are you actually trying to produce here?
> > That d1 and d2 have different time dim tags? You already have that from the very beginning.
> > Or that d1.size_placeholder is tagged but d2.size_placeholder is not tagged?
> > What case exactly? (Don't just answer "the error I have". Say _what error_, _what situation exactly_.)
> 
> I reproduce the test s.t. two Data instances have different tags for the time dim. ...

You are not answering my question. I gave multiple options in my question:
1. That d1 and d2 have different time dim tags?
2. Or that d1.size_placeholder is tagged but d2.size_placeholder is not tagged?
3. Something else.

Please make it simple. Just directly answer my question (sentence by sentence, direct answer). So, do I interpret it correctly, it is case 1?